src/bin/pgcopydb/ld_replay.c

```diff
@@ -249,3 +250,136 @@ stream_replay_line(void *ctx, const char *line, bool *stop)
 
 	return true;
 }
+
+
+/* streamReplayLineIntoBuffer is a callback function for the ReadFromStreamContext
+ * and read_from_stream infrastructure. It's called on each line read from a
+ * stream such as a unix pipe and buffers the line into a file until we encounter
+ * a COMMIT. This ensures that we can replay the whole transaction in one go
+ * and avoids deadlock when the BEGIN messages arrives before the existence of
+ * transation metadata file which is created by the transform process.
+ */
+static bool
+streamReplayLineIntoBuffer(void *ctx, const char *line, bool *stop)
+{
+	ReplayStreamCtx *replayCtx = (ReplayStreamCtx *) ctx;
+	StreamApplyContext *context = &(replayCtx->applyContext);
+
+	if (strstr(line, OUTPUT_BEGIN) == line)
+	{
+		/* If there is already a replay buffer open, it indicates a partial
+		 * transaction. Close it and start a new one.
+		 */
+		if (replayCtx->txnReplayBuffer != NULL)
+		{
+			fclose(replayCtx->txnReplayBuffer);
+		}
+
+		LogicalMessageMetadata metadata = { 0 };
+		if (!parseSQLAction((char *) line, &metadata))
+		{
+			/* errors have already been logged */
+			return false;
+		}
+
+		/* Create a txid as a symlink to the buffer file */
+		char txidFile[MAXPGPATH];
+		sformat(txidFile, MAXPGPATH, "%s/%d.sql", context->paths.dir, metadata.xid);
+
+		/* Open replaybuffer in w+ mode to truncate if the file already exists */
+		replayCtx->txnReplayBuffer = fopen_with_umask(txidFile,
+													  "w+", O_RDWR | O_TRUNC | O_CREAT,
+													  0644);
+		if (replayCtx->txnReplayBuffer == NULL)
+		{
+			log_error("Failed to open transaction buffer file \"%s\": %m",
+					  txidFile);
+			return false;
+		}
+
+		if (!create_symbolic_link(txidFile, context->paths.latesttxnfile))
+		{
+			/* errors have already been logged */
+			return false;
+		}
+
+		fformat(replayCtx->txnReplayBuffer, "%s\n", line);
+	}
+	else if (strstr(line, OUTPUT_COMMIT) == line)
+	{
+		if (replayCtx->txnReplayBuffer == NULL)
+		{
+			log_debug("Ignore COMMIT message received without a preceding BEGIN");
+			return true;
+		}
```

Comment:
Same question, does that ever happen without being an ERROR or at least a WARNing? when does that situation occur in a normal execution?